### ISSUE TITLE
New

### ISSUE THREAD

**Issue by killerhriday** [killerhriday/Hriday-Patel-stock-market#1]
import dash
from dash import dcc, html, Input, Output
from dash.exceptions import PreventUpdate
import flask
import pandas as pd
import yfinance as yf
from datetime import date, timedelta, datetime
import plotly.graph_objects as go
from plotly.subplots import make_subplots
from sklearn.linear_model import LinearRegression
import numpy as np
import os

# Clear the log file at the beginning of the script
with open("app.log", "w") as f:
    pass  # This line clears the contents of the file

app = dash.Dash(__name__)

active_graph = "stock"  # Initialize the active graph type variable

# Function to log the stock input and time
def log_stock_input(stock_symbol):
    with open("app.log", "a") as f:
        timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        f.write(f"{timestamp}: Stock symbol entered: {stock_symbol}\n")

def get_stock_type(stock_info):
    exchange = stock_info.info['exchange']
    if exchange == 'NYQ':
        return 'NYSE'
    elif exchange == 'NMS':
        return 'NASDAQ'
    else:
        return exchange

def generate_candlestick_chart(stock_data, stock_name, input_string, stock_type, graph_height):
    fig = make_subplots(rows=1, cols=1)

    X = np.array(range(1, 6)).reshape(-1, 1)
    y = stock_data['Close'].tail(5).values

    model = LinearRegression()
    model.fit(X, y)

    next_day_index = np.array([6]).reshape(-1, 1)
    next_day_price = model.predict(next_day_index)[0]

    candlestick_trace = go.Candlestick(x=stock_data["Date"],
                                       open=stock_data["Open"],
                                       high=stock_data["High"],
                                       low=stock_data["Low"],
                                       close=stock_data["Close"])
    fig.add_trace(candlestick_trace)

    ma_windows = [7, 20, 50, 100]
    for window in ma_windows:
        ma_trace = go.Scatter(x=stock_data["Date"], y=stock_data['Close'].rolling(window=window).mean(),
                              mode='lines', line=dict(width=2), name=f'{window}-days MA')
        fig.add_trace(ma_trace)

    fig.add_trace(go.Scatter(x=[stock_data["Date"].iloc[-1] + timedelta(days=1)], y=[next_day_price],
                             mode='markers',
                             marker=dict(symbol='circle', color='red', size=5),
                             name='Predicted Price'), row=1, col=1)

    # Adjust x-axis range to include a buffer zone beyond the last data point
    x_min = min(stock_data["Date"])
    x_max = max(stock_data["Date"]) + timedelta(days=1)  # Add one day to extend beyond the last data point
    x_buffer = (x_max - x_min) * 0.01  # Define buffer as 1% of the x-axis range
    x_axis_range = [x_min - x_buffer, x_max + x_buffer]

    fig.update_layout(title_text=f"Stock Price Analysis - {stock_name} ({input_string}) ({stock_type})",
                      xaxis_rangeslider_visible=False,
                      height=graph_height,
                      xaxis=dict(title='Date', tickfont=dict(size=12), showline=True, linewidth=1, linecolor='black', range=x_axis_range),
                      yaxis=dict(title='Price', tickfont=dict(size=12), showline=True, linewidth=1, linecolor='black'),
                      legend=dict(font=dict(size=12)),
                      plot_bgcolor='rgba(0, 0, 255, 0.05)',
                      paper_bgcolor='rgba(0, 0, 255, 0.05)')

    return fig, stock_data

def generate_volume_chart(stock_data, stock_name, input_string, stock_type, graph_height):
    volume_fig = go.Figure()
    volume_fig.add_trace(go.Bar(x=stock_data["Date"], y=stock_data["Volume"], marker=dict(color='blue'), name="Volume"))
    volume_fig.update_layout(title_text=f"Stock Volume Analysis - {stock_name} ({input_string}) ({stock_type})",
                             xaxis=dict(title='Date', tickfont=dict(size=12), showline=True, linewidth=1,
                                        linecolor='black'),
                             yaxis=dict(title='Volume', tickfont=dict(size=12), showline=True, linewidth=1,
                                        linecolor='black'),
                             plot_bgcolor='rgba(0, 0, 255, 0.05)',
                             paper_bgcolor='rgba(0, 0, 255, 0.05)',
                             height=graph_height)  # Set the height to match the candlestick graph
    return volume_fig

input_style = {
    'listStyleType': 'none',
    'margin': '0',
    'padding': '0',
    'overflow': 'hidden',
    'backgroundColor': 'rgb(124, 130, 178)'
}

input_item_style = {
    'float': 'left',
}

input_link_style = {
    'display': 'block',
    'color': 'white',
    'textAlign': 'center',
    'padding': '14px 16px',
    'textDecoration': 'none',
}

app.layout = html.Div(
    style={'backgroundColor': 'rgb(208, 210, 234)', 'height': 'auto', 'overflow': 'hidden'},
    children=[
        html.H1("Stock Price Analysis",
                style={'textAlign': 'center', 'marginBottom': '2vh', 'fontSize': '3vw', 'fontfamily':'Times New Roman'}),
        html.Div(id='input-container', children=[
            html.Ul(style=input_style, children=[
                html.Li(style=input_item_style, children=[
                    html.A(dcc.Input(id='stock-input', type='text', placeholder='Enter stock symbol...',
                                      debounce=True, style={'marginRight': '1vw'}),
                           href="#", n_clicks=0, style=input_link_style)
                ]),
                html.Li(style=input_item_style, children=[
                    html.A('Stocks Graph', id='submit-button', href="#", n_clicks=0, style=input_link_style)
                ]),
                html.Li(style=input_item_style, children=[
                    html.A('Volume Graph', id='volume-button', href="#", n_clicks=0, style=input_link_style)
                ]),
            ]),
        ]),
        html.Div(id='date-range-container', style={'display': 'none', 'alignItems': 'center'}, children=[
            dcc.DatePickerRange(
                id='date-range-picker',
                display_format='DD/MM/YYYY',
                max_date_allowed=date.today(),  # Disables future dates
                start_date_placeholder_text="Start Date",
                end_date_placeholder_text="End Date",
                style={'marginRight': '1vw'}
            ),
            html.Div(style={'marginLeft': '1vw'}, children=[
                html.Button('Update', id='update-button', n_clicks=0,
                            style={'backgroundColor': '#4CAF50',
                                   'color': 'white', 'border': 'none', 'padding': '1vh 2vw',
                                   'text-align': 'center', 'text-decoration': 'none',
                                   'display': 'inline-block', 'fontSize': '1.5vw',
                                   'borderRadius': '1vw', 'cursor': 'pointer'}),
                html.Button('Reset', id='reset-button', n_clicks=0,
                            style={'backgroundColor': '#f44336', 'color': 'white', 'border': 'none',
                                   'padding': '1vh 2vw', 'text-align': 'center',
                                   'text-decoration': 'none', 'display': 'inline-block',
                                   'fontSize': '1.5vw', 'borderRadius': '1vw', 'cursor': 'pointer',
                                   'marginLeft': '1vw'}),
            ]),
        ]),
        html.Div(id='output-container', style={'display': 'flex', 'backgroundColor': 'rgb(208, 210, 232)'}),
        html.Div(id='output-graph', style={'bold': '1', 'backgroundColor': 'rgb(208, 210, 232)'})
    ])

@app.server.route('/download-data/<path:path>')
def download_data(path):
    return flask.send_file(path, as_attachment=True)

@app.callback(
    [Output('output-graph', 'children'),
     Output('date-range-container', 'style'),
     Output('date-range-picker', 'start_date'),
     Output('date-range-picker', 'end_date')],
    [Input('submit-button', 'n_clicks'),
     Input('volume-button', 'n_clicks'),
     Input('reset-button', 'n_clicks'),
     Input('update-button', 'n_clicks')],
    [dash.dependencies.State('stock-input', 'value'),
     dash.dependencies.State('date-range-picker', 'start_date'),
     dash.dependencies.State('date-range-picker', 'end_date')]
)
def update_graph(submit_clicks, volume_clicks, reset_clicks, update_clicks, input_string, start_date, end_date):
    global active_graph  # Declare the active graph variable as global

    ctx = dash.callback_context
    if not ctx.triggered:
        raise PreventUpdate

    if ctx.triggered[0]['prop_id'] == 'reset-button.n_clicks':
        if active_graph == "stock":
            today = date.today()
            end_date = today.strftime("%Y-%m-%d")
            start_date = (today - timedelta(days=500)).strftime("%Y-%m-%d")
        elif active_graph == "volume":
            today = date.today()
            end_date = today.strftime("%Y-%m-%d")
            start_date = (today - timedelta(days=500)).strftime("%Y-%m-%d")

    # Update active graph based on which button is clicked
    if submit_clicks or volume_clicks:
        if ctx.triggered[0]['prop_id'] == 'submit-button.n_clicks':
            active_graph = "stock"
            log_stock_input(input_string)  # Log the stock input
        elif ctx.triggered[0]['prop_id'] == 'volume-button.n_clicks':
            active_graph = "volume"
            with open("app.log", "a") as f:
                timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
                f.write(f"{timestamp}: Volume graph opened for stock symbol: {input_string}\n")

    stock_info = yf.Ticker(input_string)
    stock_name = stock_info.info['longName']
    stock_type = get_stock_type(stock_info)

    if not start_date or not end_date:
        today = date.today()
        end_date = today.strftime("%Y-%m-%d")
        start_date = (today - timedelta(days=500)).strftime("%Y-%m-%d")

    data = yf.download(input_string, start=start_date, end=end_date, progress=False)
    data.reset_index(inplace=True)

    # Calculate graph height based on screen height
    screen_height = flask.request.headers.get('User-Agent').lower()
    if "mobile" in screen_height or "android" in screen_height or "iphone" in screen_height:
        graph_height = 400  # Height for mobile devices
    elif "screen" in screen_height and "height" in screen_height:
        try:
            height_str = screen_height.split("height=")[1].split(",")[0]
            height = int(height_str)
            if height < 800:
                graph_height = height - 100  # Adjusting height to fit smaller screens
            else:
                graph_height = 800
        except:
            graph_height = 800  # Default height if parsing fails
    else:
        graph_height = 800  # Height for other devices

    if active_graph == "stock":
        candlestick_chart, stock_data = generate_candlestick_chart(data, stock_name, input_string, stock_type,
                                                                  graph_height)
        # Apply time range to stock chart
        candlestick_chart.update_layout(xaxis_range=[start_date, end_date])
        graph = dcc.Graph(figure=candlestick_chart)
    elif active_graph == "volume":
        volume_chart = generate_volume_chart(data, stock_name, input_string, stock_type, graph_height)
        # Apply time range to volume chart
        volume_chart.update_layout(xaxis_range=[start_date, end_date])
        graph = dcc.Graph(figure=volume_chart)

    # Log the graph type
    with open("app.log", "a") as f:
        timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        if active_graph == "stock":
            f.write(f"{timestamp}: Stock graph opened for stock symbol: {input_string}\n")
        elif active_graph == "volume":
            f.write(f"{timestamp}: Volume graph opened for stock symbol: {input_string}\n")

    return graph, {'display': 'block'}, start_date, end_date

if __name__ == '__main__':
    app.run_server(host='0.0.0.0', port=19001, debug=True)
